### DISCUSSION TITLE
database disk image is malformed?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#5541] [Q&A]
I'm seeing these errors occasionally on PocketHost and was wondering if there might be an `sqlite3` command I can run on behalf of users periodically to clean up things like this:

```
error: 2024/09/17 20:47:14 Failed to write log &{{false pxs2f6tfa0prfrx 2024-09-17 20:47:11.518Z 2024-09-17 20:47:11.518Z} 
map[
  auth:guest 
  execTime:11.449576 
  method:GET 
  referer: remoteIp:******* 
  status:200 
  type:request 
  url:/api/collections/countries/records?page=1&perPage=500&skipTotal=1 
  userAgent:node userIp:********
] 
GET /api/collections/countries/records?page=1&perPage=500&skipTotal=1 0} 

database disk image is malformed (11)
```

**Comment by ganigeorgiev** [comment]
I'm not aware of `sqlite3` command that will autofix such error.

Specifically for the logs, you can delete the logs.db file and restart the application (it should automatically recreate it on startup).

But in general `database disk image is malformed` is not something that you should ignore and I recommend to try to investigate why it happens in the first place.

The scenarios where I could think of that could happen are:

1. The user has partially copied their local `pb_data` files to the remote (e.g. skipped the `logs.db-wal` file if such exist).
2. The user manually copied their local `pb_data` files to the remote while there was a log write in the background (since the `logs.db` and `logs.db-wal` will be copied at slightly different times, their content may not be compatible with each other).
3. Restore from the builtin backups. This is difficult to reproduce but I guess it is possible to happen in case there is a delayed logs batch write during backup generation. I've submitted a fix to explicitly lock the logs database in addition to the `data.db` in [v0.22.21](https://github.com/pocketbase/pocketbase/releases/tag/v0.22.21).
    _(To be honest I'm not sure if logs need to be part of the backup in the first place but for now let's keep it as it is and I'll revisit it in the future together with https://github.com/pocketbase/pocketbase/issues/4647.)_
4. Conflicting third party application that interacts directly with the database or the PocketBase executable while the app server is running (_the latter could happen because the console commands execute in their own separate process_).

If the error in your case is not caused by manual user copying then I suspect that it is most likely the 3rd case but it could be also something totally different, I'm not sure. Try to upgrade and monitor it for a while and if it still happens try to pinpoint when it happens and what has triggered it.

**Reply by benallfree** [reply]
Thanks for the ideas.

(1) seems very likely

(2) rare, but possible. PocketHost only allows writing to pb_data when the instance is off, but there could be race conditions

(3) Maybe. Users do occasionally struggle with bad restores in other situations. For example, today a user's restore conflicted with existing pb_migrations (they performed the same schama change on local and pockethost prior to the restore), thus causing a conflict. Maybe something similar is happening with logs?

(4) This is not allowed on PocketHost, so probably not. 

So far the issue seems isolated to specific instances so I suspect there might be some user error involved. I think it would be more widespread if it had something to do with improperly shutting down instances or something like that.

**Reply by ganigeorgiev** [reply]
>  For example, today a user's restore conflicted with existing pb_migrations (they performed the same schama change on local and pockethost prior to the restore), thus causing a conflict. Maybe something similar is happening with logs?

No, I don't think this is the case because the logs schema rarely have to change (especially by users).

In general collection schema changes between environments should be done either via migrations or via the export/import UI but the 2 options shouldn't be mixed because at the minimum the ids of the created collections/fields could differ between the environments if applied in such way. With the next release the autogenerated ids of the collections and their fields will be crc32 checksum of the collection/field `name` instead of random generated string that should minimize to some extend such conflicts if the different manual changes are seemingly the same between the environments but this doesn't cover all cases and I don't think there is an easy way to prevent conflicts when there are multiple sources of truth for the schema state.
